TokenChecker.php

```diff
@@ -0,0 +1,145 @@
+<?php 
+
+require_once "phpseclib/Crypt/RSA.php";
+require_once "EndpointHandler.php";
+require_once "settings.php";
+
+	
+// A class to verify an id_token, following the implicit flow
+// defined by the OpenID Connect standard
+class TokenChecker {
+	
+	// Class variables
+	private $id_token_array = array();
+	private $head = "";
+	private $payload = "";
+	private $clientID = "";
+	private $endpointHandler; 
+	
+	function __construct($id_token, $clientID, $policy_name) {
+		
+		$this->clientID = $clientID;
+		$this->endpointHandler = new EndpointHandler($policy_name);
+		$this->splitIdToken($id_token);
+	}
+	
+	// Converts base64url encoded string into base64 encoded string
+	// Also adds the necessary padding to the base64 encoded string
+	private function convert_base64url_to_base64($data) {
+		return str_pad(strtr($data, '-_', '+/'), strlen($data) % 4, '=', STR_PAD_RIGHT);
+	}
+	
+	// Splits the id token into an array of header, payload, and signature
+	private function splitIdToken($id_token) {
+
+		// Split the token into Header, Payload, and Signature, and decode
+		$this->id_token_array = explode('.', $id_token);
+		$this->head = base64_decode($this->id_token_array[0]);
+		$this->payload = base64_decode($this->id_token_array[1]);
+	}
+	
+	// Validates the RSA signature on the token
+	private function validateSignature() {
+		
+		// Get kid from header
+		$kid = getClaim("kid", $this->head);
+		
+		// For each JwksURI, get the public key and verify the signature
+		$key_datas = $this->endpointHandler->getJwksUriData();
+		foreach ($key_datas as $key_data) {
+			
+			// Iterate through each key type until the one that matches the "kid" is found
+			$keys = json_decode($key_data, true)["keys"];
+			foreach ($keys as $key) {
+				
+				if ($key["kid"] == $kid) {
+					$e = $key["e"];
+					$n = $key["n"];
+					break; 
+				}
+			}
+			
+			// 'e' and 'n' are base64 URL encoded, change to just base64 encoding
+			$e = $this->convert_base64url_to_base64($e);
+			$n = $this->convert_base64url_to_base64($n);
+
+			// Convert RSA(e,n) format to PEM format
+			$rsa = new Crypt_RSA();
+			$rsa->setPublicKey('<RSAKeyValue>
+				<Modulus>' . $n . '</Modulus>
+				<Exponent>' . $e . '</Exponent>
+				</RSAKeyValue>');
+			$public_key = $rsa->getPublicKey();
+			
+			// Construct data and signature for verification
+			$to_verify_data = $this->id_token_array[0] . "." . $this->id_token_array[1];
+			$to_verify_sig = base64_decode($this->convert_base64url_to_base64(($this->id_token_array[2])));
+			
+			// Get hash type for verification
+			$alg = getClaim("alg", $this->head);
+			switch($alg)  {
+				case 'RS256':
+					$verified = openssl_verify($to_verify_data, $to_verify_sig, $public_key, OPENSSL_ALGO_SHA256);
```

Comment:
In response to [https://github.com/AzureAD/active-directory-b2c-wordpress-plugin-openidconnect/pull/1#discussion_r70005064](https://github.com/AzureAD/active-directory-b2c-wordpress-plugin-openidconnect/pull/1#discussion_r70005064):

> I originally looked into the libraries at jwt.io but found they required the signing key in the constructor. However, to get the key, I had to decode the token and get the kid first.

The `JWT::decode` method in [`firebase/php-jwt`](https://github.com/firebase/php-jwt) accepts an associative array of JWKs (where the array's key is the `kid` value, and the value is the PEM-encoded certificate). If the JWT has a `kid` value and you provide an array of keys, it will pick the corresponding one: [https://github.com/firebase/php-jwt/blob/master/src/JWT.php#L100](https://github.com/firebase/php-jwt/blob/master/src/JWT.php#L100).

[`SpomkyLabs/jose`](https://github.com/Spomky-Labs/jose) has a that accepts a JWK Set (i.e. all the keys). (It doesn't use the `kid` attribute, however, it just tries against all of them.) [https://github.com/Spomky-Labs/jose/blob/master/src/Verifier.php#L70](https://github.com/Spomky-Labs/jose/blob/master/src/Verifier.php#L70)